lib/uplink/main.go

```diff
@@ -0,0 +1,238 @@
+type Uplink struct {
+}
+
+// NewUplink creates a new Uplink
+func NewUplink(identity Identity, satelliteAddr string, cfg Config) *Uplink {
+	panic("TODO")
+}
+
+// A Share is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+type Share struct {
+	Access Macaroon
+
+	// TODO: these should be per-bucket somehow maybe? oh man what a nightmare
+	PathCipher       storj.Cipher
+	EncPathPrefix    storj.Path
+	Key              storj.Key
+	EncryptionScheme storj.EncryptionScheme
+}
+
+// ParseShare parses a serialized Share
+func ParseShare(data []byte) (Share, error) {
+	panic("TODO")
+}
+
+func (s *Share) Serialize() ([]byte, error) {
+	panic("TODO")
+}
+
+// Session represents a specific access session.
+type Session struct {
+}
+
+// A Session is created with a Share.
+func (u *Uplink) Session(share Share) *Session {
+	panic("TODO")
+}
+
+// GetBucket returns info about the requested bucket if authorized
+func (s *Session) GetBucket(ctx context.Context, bucket string) (storj.Bucket,
+	error) {
+	panic("TODO")
+}
+
+type CreateBucketOptions struct {
+	PathCipher Cipher
+	// this differs from storj.CreateBucket's choice of just using storj.Bucket
+	// by not having 2/3 unsettable fields.
+}
+
+// CreateBucket creates a new bucket if authorized
+func (s *Session) CreateBucket(ctx context.Context, bucket string,
+	opts *CreateBucketOptions) (storj.Bucket, error) {
+	panic("TODO")
+}
+
+// DeleteBucket deletes a bucket if authorized
+func (s *Session) DeleteBucket(ctx context.Context, bucket string) error {
+	panic("TODO")
+}
+
+// ListBuckets will list authorized buckets
+func (s *Session) ListBuckets(ctx context.Context, opts storj.BucketListOptions) (
+	storj.BucketList, error) {
+	panic("TODO")
+}
+
+// Share creates a new share, potentially further restricted from the Share used
+// to create this session.
+func (s *Session) Share(ctx context.Context, caveats ...Caveat) (Share, error) {
+	panic("TODO")
+}
+
+// ObjectMeta represents metadata about a specific Object
+type ObjectMeta struct {
+	Bucket   string
+	Path     storj.Path
+	IsPrefix bool
+
+	Metadata map[string]string
+
+	Created  time.Time
+	Modified time.Time
+	Expires  time.Time
+
+	Size     int64
+	Checksum string
+
+	// this differs from storj.Object by not having Version (yet), and not
+	// having a Stream embedded. I'm also not sold on splitting ContentType out
+	// from Metadata but shrugemoji.
+}
+
+// GetObject returns a handle to the data for an object and its metadata, if
+// authorized.
+func (s *Session) GetObject(ctx context.Context, bucket string, path storj.Path) (
+	ranger.Ranger, ObjectMeta, error) {
```

Comment:
`ranger` here isn't necessary using `io.ReadSeeker` would keep code clear on both sides.

Reply:
rangers and readseekers are for the most part pretty interchangeable - you can implement one with the other - but there is one major performance difference. rangers know up front how much data you're going to request. when you say GetRange, you not only indicate the starting position, but also how long the stream will be, which you don't get with ReadSeekers. this one property simplifies a bunch of little details all through the pipeline. it makes it so you can request a range which will kick off a bunch of machinery to start streaming that full range to you, which you then get a reader for. with a ReadSeeker, you may pay a penalty of doing extra pipelined work speculatively because you're not sure the caller has gotten everything they need

Reply:
The goal of `uplink` as a package is to make things as simple as possible for common use-cases. This means upload a small full file and download full file. If you want something more then you would go to the other packages.

As for ReadSeeker, yes there would be some amount of overhead by using it directly. Although we can probably provide some utilities that get rid of that as well (e.g. `ReadSeekLimited`).

Need to figure out how large the additional cost is.

Reply:
:+1: